File: lib/features/auto-typecheck.js
'use babel'

import SubAtom from 'sub-atom'

export default class AutoTypecheck {
  constructor (editor, clientLookup) {
    this.editor = editor
    this.clientLookup = clientLookup
    this.disposables = new SubAtom()

    const buffer = this.editor.getBuffer()
    this.disposables.add(buffer.onDidSave(() => {
      // typecheck file on save
      let needle
      if ((needle = atom.config.get('Ensime.typecheckWhen'), ['save', 'typing'].includes(needle))) {
        return __guard__(this.clientLookup(), x => x.typecheckFile(this.editor.getPath()))
      }
    })
    )

    // Typecheck buffer while typing
    this.disposables.add(atom.config.observe('Ensime.typecheckWhen', value => {
      if (value === 'typing') {
        this.typecheckWhileTypingDisposable = this.editor.onDidStopChanging(() => {
          const b = this.editor.getBuffer()
          const p = b.getPath()
          // Don't typecheck deps
          if (!p.includes('dep-src')) {
            return __guard__(this.clientLookup(), x => x.typecheckBuffer(p, b.getText()))
          }
        })
        return this.disposables.add(this.typecheckWhileTypingDisposable)
      } else {
        this.disposables.remove(this.typecheckWhileTypingDisposable)
        return (this.typecheckWhileTypingDisposable != null ? this.typecheckWhileTypingDisposable.dispose() : undefined)
      }
    })
    )
  }

  deactivate () {
    return this.disposables.dispose()
  }
}

function __guard__ (value, transform) {
  return (typeof value !== 'undefined' && value !== null) ? transform(value) : undefined
}

File: src/main/scala/features/auto-typecheck.scala
package org.ensime.atom.features

import scala.scalajs.js
import scala.scalajs.js.annotation._

import org.ensime.atom.external.{
  AtomEnvironment,
  Disposable,
  Ensime,
  LogLevel,
  SubAtom,
  TextBuffer,
  TextEditor
}

@JSExportAll
@JSExportTopLevel("AutoTypecheck")
class AutoTypecheck(editor: TextEditor,  clientLookup: js.Function0[Ensime]) {
  val disposables = new SubAtom();

  val buffer = editor.getBuffer;

  var typecheckWhileTypingDisposable: Disposable = null
  disposables.add(buffer.onDidSave(() => {
    // typecheck file on save
    if (Seq("save", "typing").contains(AtomEnvironment.config.get("Ensime.typecheckWhen")))
    if (clientLookup() != null) {
      clientLookup().typecheckFile(editor.getPath);
    }}));

   // Typecheck buffer while typing
  disposables.add(AtomEnvironment.config.observe[String]("Ensime.typecheckWhen", (value) => {
    if(value == "typing"){
      typecheckWhileTypingDisposable = editor.onDidStopChanging(() => {
        val b = editor.getBuffer;
        val p = b.getPath;
        // Don't typecheck deps
        if(!(p.contains("dep-src")) && clientLookup() != null){
          clientLookup().typecheckBuffer(p, b.getText);
        }
        disposables.add(typecheckWhileTypingDisposable);
      });
    } else {
        disposables.remove(typecheckWhileTypingDisposable);
        if (typecheckWhileTypingDisposable != null) {
          typecheckWhileTypingDisposable.dispose();
        }}
  }));

  def deactivate(): Unit = {
    disposables.dispose();
  }
};
